### PR TITLE
Replicate reference tables to the coordinator

### PR DESCRIPTION
**Pull request by pykello**
After user designates the coordinator by doing a `master_add_node(hostname, port, groupId => 0)`, we will replicate all reference tables to the coordinator too. This is so users can join reference tables with local tables too.

For queries which are a SELECT query which joins reference tables and local tables, we replace the reference table with the shard table and do local planning.

To test this, we added some new tests, but also run some related tests again after adding a coordinator node.

### CODE REVIEW

File: configure.in [hunk 1]
```diff
@@ -162,7 +162,9 @@ CITUSAC_PROG_CC_CFLAGS_OPT([-Wno-unused-parameter])
 CITUSAC_PROG_CC_CFLAGS_OPT([-Wno-sign-compare])
 CITUSAC_PROG_CC_CFLAGS_OPT([-Wno-missing-field-initializers])
 CITUSAC_PROG_CC_CFLAGS_OPT([-Wno-clobbered])
-CITUSAC_PROG_CC_CFLAGS_OPT([-Wno-gnu-variable-sized-type-not-at-end])
+if test x"$GCC" = x"no" ; then
```
Comment (pykello) [inline configure.in]: I'm not sure if this is needed anymore after Jelte's recent changes for flag check. If it is needed, we need to move it to a separate PR.

File: src/backend/distributed/utils/reference_table_utils.c [hunk 2]
```diff
@@ -389,10 +389,9 @@ uint32
...
 {
 	uint32 colocationId = INVALID_COLOCATION_ID;
-	List *workerNodeList = ActivePrimaryNodeList(ShareLock);
 	int shardCount = 1;
-	int replicationFactor = list_length(workerNodeList);
 	Oid distributionColumnType = InvalidOid;
+	int replicationFactor = ReferenceTableReplicationFactor();
 
 	/* check for existing colocations */
 	colocationId = ColocationId(shardCount, replicationFactor, distributionColumnType);
```
[the remark this answers is not in the document]
Comment (marcocitus) [reply src/backend/distributed/utils/reference_table_utils.c]: (Optional)

[the remark this answers is not in the document]
Comment (pykello) [reply src/backend/distributed/utils/reference_table_utils.c]: https://github.com/citusdata/citus/pull/3152

File: src/backend/distributed/planner/distributed_planner.c [hunk 3]
```diff
@@ -594,12 +717,14 @@ CreateDistributedPlannedStmt(uint64 planId, PlannedStmt *localPlan, Query *origi
 static DistributedPlan *
 CreateDistributedPlan(uint64 planId, Query *originalQuery, Query *query, ParamListInfo
```
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/planner/distributed_planner.c]: Thinking on this more:

There might be some tricky edge cases of not creating distributed plans and not using distributed execution (in this patch and for my above proposal as well), such as the following:

```SQL
BEGIN;

-- does a Postgres planning/execution, skipping Citus altogether
-- !!!! When asserts enabled, this crashes !!!!
SELECT count(*) FROM ref_table JOIN local_table ON (true);

-- does a distributed planning, which yields a distributed execution (e.g., connect to itself for the coordinator as well)
-- so a self deadlock
TRUNCATE ref_table;
```

Also, since we're skipping the distributed planning/execution, we might not be acquiring the correct shard locks etc. Even, `citus.log_remote_commands` wouldn't should show the query. Or, anything that Citus planner/executors does, that is relevant to the reference tables would be skipped. 

So, maybe we should re-architect the solution such that Citus is always in the picture, just conveying the plan to the local execution. In that case, we wouldn't need to consider one more edge case of queries on reference tables skipping the distributed planning/execution.

To be more specific, what do you think of treating local/ref joins as router plans? In that case, we can avoid any rough edges with this new change, (and later focus on improving the local execution). 

I guess, it shouldn't be too hard to refactor router planner to accept local tables. I think I'm even OK to have a new case, similar to `FastPathRouterPlanner()`, where we have `ReferenceTableLocalTableJoinRouterPlanner()`, which does the trick in `RouterPlanner()`.

I'm curious of your thoughts (and also @marcocitus if he's around).

File: src/backend/distributed/planner/distributed_planner.c [hunk 4]
```diff
@@ -243,6 +254,113 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
...
+	return allReferenceTables && currentNodeKnown && parse->commandType == CMD_SELECT;
+}
+
+
+/*
+ * PlanLocalReferenceTableSelect returns a plan for the given query that can be executed
+ * locally, assuming it satisfies the conditions in IsLocalReferenceTableSelect().
+ */
+static PlannedStmt *
+PlanLocalReferenceTableSelect(PlannerRestrictionContext *restrictionContext,
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c]: Note that for fast-path router queries, we do not call `standard_planner()`, which means that we don't have a valid `restrictionContext `

File: src/backend/distributed/planner/distributed_planner.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (pykello) [reply src/backend/distributed/planner/distributed_planner.c]: Overall this makes sense. I am splitting this PR into two, one which contains everything except planner related changes (https://github.com/citusdata/citus/pull/3155), and second will be planner related changes. I am sending a calendar invite for tomorrow to chat about planner changes.

**Comment by marcocitus** [comment]
Superseded by #3155
